### PR TITLE
[SPARK-33614][SQL] Fix the constant folding rule to skip it if the expression fails to execute

### PR DESCRIPTION
**Pull request by luluorta**
### What changes were proposed in this pull request?
This PR suppresses constant folding for expressions which throw exceptions (mostly in ansi mode).

### Why are the changes needed?
This PR makes the queries that have invalid expressions in unreachable branches can be executed successfully.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
added UT + existing UTs

### CODE REVIEW

**Comment by luluorta** [comment]
cc  @cloud-fan @leanken

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -61,8 +62,18 @@ object ConstantFolding extends Rule[LogicalPlan] {
       case Size(c: CreateMap, _) if c.children.forall(hasNoSideEffect) =>
         Literal(c.children.length / 2)
 
-      // Fold expressions that are foldable.
-      case e if e.foldable => Literal.create(e.eval(EmptyRow), e.dataType)
+      // Fold expressions that are foldable. To avoid repeatedly evaluating expressions which throw
+      // exceptions, we also check whether the expression's children are all literals.
+      case e if e.childrenToEvaluate.forall(_.isInstanceOf[Literal]) && e.foldable =>
+        val constant = try {
+          Some(e.eval(EmptyRow))
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:69]: I think we should do this in `Expression.foldable`. Basically if overflowing can happen, they are not foldable.

**Comment by HyukjinKwon** [comment]
cc @tgravescs too just FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:69]: hmm, then it's a big perf regression for ANSI mode, as all math operations can't do constant folding.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: We can add it only into the expressions that can do constant folding I guess.

Comment (luluorta) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: According to @cloud-fan 's advice, I refactor this patch by using `RaiseError` as constant. Now we don't need to bother with the foldable expressions which can throw exceptions, cause they are still foldable and will be folded to `RaiseError`s.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 2]
```diff
@@ -64,38 +64,81 @@ case class PrintToStderr(child: Expression) extends UnaryExpression {
...
 
+  private def errorHashCode: Int = error match {
+    case Some(err) => 31 * err.getClass.hashCode() + err.getMessage.hashCode
+    case None => 0
+  }
+
+  override def hashCode(): Int = 31 * child.hashCode() + errorHashCode
+
+  private def errorEquals(other: Option[Throwable]): Boolean = (error, other) match {
+    case (Some(e), Some(o)) => e.getClass == o.getClass && e.getMessage == o.getMessage
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:86]: is this to skip comparing the stacktrace?

**Pull request closed without merging**